src/main/java/org/apache/sysds/runtime/matrix/data/LibMatrixReorg.java

```diff
@@ -389,10 +393,78 @@ public static MatrixBlock rev( MatrixBlock in, MatrixBlock out ) {
 		return out;
 	}
 
+	public static MatrixBlock rev(MatrixBlock in, MatrixBlock out, int k) {
+		if (k <= 1 || in.isEmptyBlock(false)) {
+			System.out.println("choosing single thread");
+			return rev(in, out); // fallback to single-threaded
+
+		}
+		final int numRows = in.getNumRows();
+		final int numCols = in.getNumColumns();
+		final boolean sparse = in.isInSparseFormat();
+
+		// Prepare output block
+		out.reset(numRows, numCols, sparse);
+
+		// Before starting threads, ensure the output sparse block is allocated!
+		if (sparse) {
+			out.allocateSparseRowsBlock(false);
+		}
+
+		// Set up thread pool
+		ExecutorService pool = CommonThreadPool.get(k);
+		try {
+			int blklen = (int) Math.ceil((double) numRows / k);
+			List<Future<?>> tasks = new ArrayList<>();
+
+			for (int i = 0; i < k; i++) {
+				final int startRow = i * blklen;
+				final int endRow = Math.min((i + 1) * blklen, numRows);
+
+				tasks.add(pool.submit(() -> {
+					if (!sparse) {
+						// Dense case
+						System.out.println("dense case");
+						double[] inVals = in.getDenseBlockValues();
+						double[] outVals = out.getDenseBlockValues();
+						for (int r = startRow; r < endRow; r++) {
+							int revRow = numRows - r - 1;
+							System.arraycopy(inVals, revRow * numCols, outVals, r * numCols,
+									numCols);
+						}
+					} else {
+						// Sparse case
+						System.out.println("Sparse case");
+						SparseBlock inBlk = in.getSparseBlock();
+						SparseBlock outBlk = out.getSparseBlock();
+						for (int r = startRow; r < endRow; r++) {
+							int revRow = numRows - r - 1;
+							if (!inBlk.isEmpty(revRow)) {
+								outBlk.set(r, inBlk.get(revRow), true);
+							}
+						}
+					}
+				}));
+			}
+
+			// Wait for all threads
+			for (Future<?> task : tasks) {
+				task.get();
+			}
+		} catch (Exception ex) {
+			throw new DMLRuntimeException(ex);
+		} finally {
+			pool.shutdown();
+		}
+		out.recomputeNonZeros();
```

Comment:
avoid this unnecessary recomputation of the the number of non-zeros. We should just set the nnz of the output to the nnz of the input, as this reorganization does not change values but only placed them differently